pal/src/map/virtual.cpp

```diff
@@ -1682,6 +1673,73 @@ VirtualAlloc(
     return pRetVal;
 }
 
+#define KB64 (64 * 1024)
+#define MB64 (KB64 * 1024)
+
+LPVOID
+PALAPI
+VirtualAlloc(
+         IN LPVOID lpAddress,       /* Region to reserve or commit */
+         IN SIZE_T dwSize,          /* Size of Region */
+         IN DWORD flAllocationType, /* Type of allocation */
+         IN DWORD flProtect)        /* Type of access protection */
+{
+    if (lpAddress)
+    {
+        return VirtualAlloc_(lpAddress, dwSize, flAllocationType, flProtect);
+    }
+
+    bool reserve = (flAllocationType & MEM_RESERVE) == MEM_RESERVE;
+    bool commit  = (flAllocationType & MEM_COMMIT)  == MEM_COMMIT;
+
+    if (reserve || commit)
+    {
+        char *address = (char*) VirtualAlloc_(nullptr, dwSize, MEM_RESERVE, flProtect);
+        if (!address) return nullptr;
+
+        if (reserve)
+        {
+            flAllocationType &= ~MEM_RESERVE;
+        }
+
+        SIZE_T diff = ((ULONG_PTR)address % KB64);
+        if ( diff != 0 )
+        {
+            VirtualFree(address, 0, MEM_RELEASE);
+            char *addr64 = address + (KB64 - diff);
+
+            // try reserving from the same address space
+            address = (char*) VirtualAlloc_(addr64, dwSize, MEM_RESERVE, flProtect);
+
+            if (!address)
+            {   // looks like ``pushed new address + dwSize`` is not available
+                // try on a bigger surface
+                address = (char*) VirtualAlloc_(nullptr, dwSize + KB64, MEM_RESERVE, flProtect);
+                if (!address) return nullptr;
+
+                diff = ((ULONG_PTR)address % KB64);
+                addr64 = address + (KB64 - diff);
+
+                CPalThread *pthrCurrent = InternalGetCurrentThread();
+                InternalEnterCriticalSection(pthrCurrent, &virtual_realloc);
```

Comment:
More a question than a comment- what is the intention behind this critical section? For example, why is this not entered when you first call VirtualAlloc_(add64)